### PR TITLE
Lazy tests added #477

### PR DESCRIPTION
**Pull request by ruslansennov**

### CODE REVIEW

File: src/test/java/javaslang/LazyTest.java [hunk 1]
```diff
@@ -87,6 +91,91 @@ public void shouldBeAwareOfEvaluated() {
         assertThat(lazy.isEvaluated()).isTrue();
     }
 
+    // -- get
+
+    @Test(expected = NoSuchElementException.class)
+    public void shouldThrowGetEmpty() {
+        assertThat(Lazy.empty().get()).isEqualTo(1);
```
Comment (danieldietrich) [inline src/test/java/javaslang/LazyTest.java]: dito

File: src/test/java/javaslang/LazyTest.java [hunk 2]
```diff
@@ -87,6 +91,91 @@ public void shouldBeAwareOfEvaluated() {
...
+    }
+
+    // -- filter
+
+    @Test(expected = NoSuchElementException.class)
+    public void shouldThrowFilterEmptyLazy() {
+        assertThat(Lazy.empty().filter(ignored -> true)).isEqualTo(Lazy.empty());
```
Comment (danieldietrich) [inline src/test/java/javaslang/LazyTest.java]: dito

File: src/test/java/javaslang/LazyTest.java [hunk 3]
```diff
@@ -87,6 +91,91 @@ public void shouldBeAwareOfEvaluated() {
...
+    }
+
+    @Test
+    public void shouldFilterNonEmptyLazy() {
+        assertThat(Lazy.of(() -> 1).filter(i -> true)).isEqualTo(Lazy.of(() -> 1));
+    }
+
+    @Test(expected = NoSuchElementException.class)
+    public void shouldThrowEmptyFilterNonEmptyLazy() {
+        assertThat(Lazy.of(() -> 1).filter(i -> false)).isEqualTo(Lazy.of(() -> 1));
```
Comment (danieldietrich) [inline src/test/java/javaslang/LazyTest.java]: dito

File: src/test/java/javaslang/LazyTest.java [hunk 4]
```diff
@@ -87,6 +91,91 @@ public void shouldBeAwareOfEvaluated() {
...
+    @Test
+    public void shouldNonEmptyFilterNonEmptyLazy() {
+        assertThat(Lazy.of(() -> 1).filter(i -> true)).isEqualTo(Lazy.of(() -> 1));
+    }
+
+    // -- flatten()
+
+    @Test(expected = NoSuchElementException.class)
+    public void shouldFlattenEmptyLazy() {
+        assertThat(Lazy.empty().flatten()).isEqualTo(Lazy.empty());
```
Comment (danieldietrich) [inline src/test/java/javaslang/LazyTest.java]: dito

File: src/test/java/javaslang/LazyTest.java [hunk 5]
```diff
@@ -87,6 +91,91 @@ public void shouldBeAwareOfEvaluated() {
...
+    @Test
+    public void shouldFlatMapNonEmptyLazy() {
+        assertThat(Lazy.of(() -> 1).flatMap(List::of)).isEqualTo(Lazy.of(() -> 1));
+    }
+
+    // -- map
+
+    @Test(expected = NoSuchElementException.class)
+    public void shouldMapEmpty() {
+        assertThat(Lazy.empty().map(Function.identity())).isEqualTo(Lazy.empty());
```
Comment (danieldietrich) [inline src/test/java/javaslang/LazyTest.java]: dito

File: src/test/java/javaslang/LazyTest.java [hunk 6]
```diff
@@ -87,6 +91,91 @@ public void shouldBeAwareOfEvaluated() {
...
+    public void shouldMapNonEmpty() {
+        assertThat(Lazy.of(() -> 1).map(i -> i + 1)).isEqualTo(Lazy.of(() -> 2));
+    }
+
+    // -- peek
+
+    @Test(expected = NoSuchElementException.class)
+    public void shouldPeekEmpty() {
+        assertThat(Lazy.empty().peek(t -> {
+        })).isEqualTo(Lazy.empty());
```
Comment (danieldietrich) [inline src/test/java/javaslang/LazyTest.java]: dito

File: src/test/java/javaslang/LazyTest.java [hunk 7]
```diff
@@ -87,6 +91,91 @@ public void shouldBeAwareOfEvaluated() {
...
+    public void shouldFlattenLazyOfLazy() {
+        assertThat(Lazy.of(Lazy.of(() -> 1)).flatten()).isEqualTo(Lazy.of(() -> 1));
+    }
+
+    // -- flatMap
+
+    @Test(expected = NoSuchElementException.class)
+    public void shouldFlatMapEmptyLazy() {
+        assertThat(Lazy.empty().flatMap(List::of)).isEqualTo(Lazy.empty());
+    }
```
Comment (danieldietrich) [inline src/test/java/javaslang/LazyTest.java]: dito

**Comment by danieldietrich** [comment]
thx for adding so many tests - that really helps!
I need a day to implement the RedBlackTree to make progress towards 2.0.0. Our goal comes closer...

**Comment by ruslansennov** [comment]
> Our goal comes closer...

We need to implement most of <code>HashMap</code> methods.
@patrox I will do this if you don't have enough time

**Comment by danieldietrich** [comment]
we should use the new `javaslang.collection.Iterator` where applicable when implementing HashMap methods...

**Comment by ruslansennov** [comment]
natürlich

**Comment by danieldietrich** [comment]
:-D

**Comment by danieldietrich** [comment]
> I need a day to implement the RedBlackTree

...this means I estimated one day of work - I hope that I can afford the time this week...

**Pull request merged**